src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,541 @@
+                  rw eval_interpolate_at_node _ hvs hi, exact dif_pos hi,
+                end }
+
+theorem interpolate_eq_sum_interpolate_insert_sdiff (hvt : set.inj_on v t) (hs : s.nonempty)
+  (hst : s ⊆ t) : interpolate t v r =
+  ∑ i in s, (interpolate (insert i (t \ s)) v r) * basis s v i :=
+begin
+  symmetry, refine eq_interpolate_of_eval_eq _ hvt  (lt_of_le_of_lt (degree_sum_le _ _) _)
+                                                    (λ i hi, _),
+  { simp_rw [(finset.sup_lt_iff (with_bot.bot_lt_coe t.card)), degree_mul], intros i hi,
+    have hs : 1 ≤ s.card := (nonempty.card_pos ⟨_, hi⟩),
+    have hst' : s.card ≤ t.card := (card_le_of_subset hst),
+    have H : t.card = (1 + (t.card - s.card)) + (s.card - 1),
+    { rw [  add_assoc, tsub_add_tsub_cancel hst' hs,
+            ← add_tsub_assoc_of_le (hs.trans hst'), nat.succ_add_sub_one, zero_add] },
+    rw [  degree_basis (set.inj_on.mono hst hvt) hi, H, with_bot.coe_add,
+          with_bot.add_lt_add_iff_right (@with_bot.coe_ne_bot _ (s.card - 1))],
+    convert (degree_interpolate_lt _ (hvt.mono (coe_subset.mpr
+            (insert_subset.mpr ⟨hst hi, sdiff_subset _ _⟩)))),
+    rw [card_insert_of_not_mem (not_mem_sdiff_of_mem_right hi), card_sdiff hst, add_comm] },
+  { simp_rw [eval_finset_sum, eval_mul], by_cases hi' : i ∈ s,
+    { rw ← add_sum_erase _ _ hi',
+      rw [eval_basis_eq (hvt.mono hst) hi'],
+      rw eval_interpolate_at_node _ (hvt.mono (coe_subset.mpr
+            (insert_subset.mpr ⟨hi, sdiff_subset _ _⟩))) (mem_insert_self _ _),
+      rw [mul_one, add_right_eq_self],
+      refine sum_eq_zero (λ j hj, _), rw mem_erase at hj, rcases hj with ⟨hij, hj⟩,
+      rw [eval_basis_ne hij hi', mul_zero] },
+    { have H : ∑ j in s, eval (v i) (basis s v j) = 1,
+      { rw ← eval_finset_sum, rw basis_sum (hvt.mono hst) hs, rw eval_one},
+      rw ←mul_one (r i), rw ← H, rw mul_sum,
+      refine sum_congr rfl (λ j hj, _), congr,
+      refine eval_interpolate_at_node _ (hvt.mono (insert_subset.mpr ⟨hst hj, sdiff_subset _ _⟩))
+                                (mem_insert .mpr (or.inr (mem_sdiff.mpr ⟨hi, hi'⟩))) } }
+end
+
+theorem interpolate_eq_add_interpolate_erase (hvs : set.inj_on v s) (hi : i ∈ s) (hj : j ∈ s)
+  (hij : i ≠ j) : interpolate s v r = interpolate (s.erase j) v r * basis_divisor (v i) (v j) +
+  interpolate (s.erase i) v r * basis_divisor (v j) (v i) :=
+begin
+  rw interpolate_eq_sum_interpolate_insert_sdiff _ hvs ⟨i, (mem_insert_self i {j})⟩ _,
+  rw sum_insert (not_mem_singleton.mpr hij), rw sum_singleton,
+  rw basis_doubleton_left hij, rw basis_doubleton_right hij, congr;
+  [skip, rw pair_comm]; ext k; rw [mem_insert, mem_erase, mem_sdiff, mem_insert, mem_singleton];
+  push_neg; exact ⟨ by { rintros (rfl | ⟨_, _, _⟩); tauto },
+                    by { rw or_iff_not_imp_left, rintros ⟨_, _⟩ _, tauto, } ⟩,
+  intros k hk, simp at hk, rcases hk with (rfl | rfl), exact hi, exact hj
+end
```

Comment:
Here's a more principled proof, but it needs the following `finset` lemmas. Could you add those to `data/finset/basic.lean`?

```lean
lemma finset.sdiff_sdiff (s t u : finset ι) :
  (s \ t) \ u = (s \ t) ∩ (s \ u) :=
by ext; finish

lemma finset.sdiff_insert (s t : finset ι) (x : ι) :
  s \ insert x t = (s \ t).erase x :=
by simp only [← sdiff_singleton_eq_erase, insert_eq, sdiff_sdiff, sdiff_union_distrib, inter_comm]

lemma finset.insert_sdiff_insert {s t : finset ι} {x : ι} (hxs : x ∈ s) (hxt : x ∉ t) :
  insert x (s \ insert x t) = s \ t :=
by rw [sdiff_insert, insert_erase (mem_sdiff.mpr ⟨hxs, hxt⟩)]

lemma finset.doubleton_comm (x y : ι) : ({x, y} : finset ι) = {y, x} :=
by ext; finish
```

```suggestion
begin
  rw [interpolate_eq_sum_interpolate_insert_sdiff _ hvs ⟨i, (mem_insert_self i {j})⟩ _,
      sum_insert (not_mem_singleton.mpr hij), sum_singleton,
      basis_doubleton_left hij, basis_doubleton_right hij,
      finset.insert_sdiff_insert hi (not_mem_singleton.mpr hij), sdiff_singleton_eq_erase,
      finset.doubleton_comm, finset.insert_sdiff_insert hj (not_mem_singleton.mpr hij.symm),
      sdiff_singleton_eq_erase],
  { exact insert_subset.mpr ⟨hi, singleton_subset_iff.mpr hj⟩ },
end
```

Reply:
I've added this. Note that the sdiff lemmas in finset, compared to the equivalent lemmas in generalized boolean algebra, seem piecemeal and a little bit of a mess in terms of naming. I haven't tried to fix that here but I just want to note that that could be a good future PR.